Globe view to mercator transition: Symbols

karimnaaji opened the pull request:
Continuing on #10984 before starting porting on the globe view gl-native branch:

Apply the same process that was done on the sphere geometry on the symbols.

I investigated first a CPU approach, where symbols were projected each in their respective projections and the interpolation was done afterwards but this ended up being an issue: We would need to evaluate symbol collision boxes at every frame at runtime. Currently, only collision circles are evaluated as such. For collision boxes, we only evaluate a few properties of symbols dynamically. Updating their positions would add a non-negligible performance and complexity overhead for something that is only applied on a particular feature (globe view) and not always (between two zoom levels).

For the reason above, I eventually landed on a GPU approach that is simpler and has less overhead with the disadvantage that symbol collision boxes aren't aligned during that phase, but it is a tradeoff that we can mitigate further by introducing a fixed duration animation.

The zoom used in the video below is only representative of the feature, not the actual zoom at which we will transition:

https://user-images.githubusercontent.com/7061573/132710685-8d5d3cb2-32aa-4979-b3b0-3251864dac47.mov

mpulkki-mapbox commented on src/shaders/_prelude.vertex.glsl:
```diff
@@ -16,12 +16,37 @@ precision highp float;
     return (w == min) ? max : w;
 }
 
+vec3 mix_globe_mercator(mat4 matrix, vec2 tile_anchor, vec3 position, vec3 tile_id, vec2 mercator_center, float t) {
+#ifdef PROJECTION_GLOBE_VIEW
+    float tiles = pow(2.0, tile_id.z);
+
+    vec2 mercator = (tile_anchor / EXTENT + tile_id.xy) / tiles;
+    mercator -= mercator_center;
+    mercator.x = wrap(mercator.x, -0.5, 0.5);
+
+    vec4 mercator_tile = vec4(mercator.xy * EXTENT, EXTENT / (2.0 * PI), 1.0);
+    mercator_tile = matrix * mercator_tile;
+
+    return mix(position, mercator_tile.xyz, vec3(t));
+#else
+    return position;
+#endif
+}
```
So if I understood right you're using the inversion matrix for taking the mercator location of the coordinate in to the encoded space? And this way you can use the same globe matrix for both points?

karimnaaji replied:
Yes exactly, this is the simplest approach that I have found. To be able to do that using the same globe projection matrix later on we need to revert the rotation + encoding so that we can put these vertices in a flat plane using that same matrix. The other approach would be to transform all of these points in that space on the CPU _and_ pass another projection per tile coordinate, but that ends being a lot of extra changes for the same result.

mpulkki-mapbox commented on src/shaders/symbol_icon.vertex.glsl:
```diff
@@ -84,10 +91,14 @@ void main() {
+        u_inv_rot_matrix, tileAnchor, vec3(a_projected_pos.xy, anchorZ),
+        u_tile_id, u_merc_center, u_zoom_transition);
+
 #ifdef PROJECTED_POS_ON_VIEWPORT
```
the `PROJECTED_POS_ON_VIEWPORT` define is used to separate coordinates that are in normalized ecef space (point labels) from coordinates in label spaces (viewport or tile). Elevation can be added only to point placed symbols as they exists in the correct space. Name of the define is a bit misleading and outdated as it's mostly about point placement vs line placement.

I don't think this will work with line labels as you're potentially interpolating between viewport coordinates and map coordinates.

![line_label_transition](https://user-images.githubusercontent.com/55925868/132740791-dac55432-c434-46d5-80b5-70c0d3c8d8a5.gif)

Do you think if we should skip those for now or find a solution?

karimnaaji replied:
Good point. Thinking about this, a potential way forward about this could be to leverage the fact that we have runtime updates of collision circle placement, so we could readjust their position by encoding the transition interpolation in `projection.projectTilePoint` (only for line labels). Investigating that a bit.

karimnaaji replied:
I have temporarily disabled transition of line place labels in https://github.com/mapbox/mapbox-gl-js/pull/10999/commits/40beb78d9714be2ad8a2c2e9c35ee2b709e01f3c to have something we can use for now, and will continue investigation of the approach I mentioned above as I think that could be something to pursue further.

mpulkki-mapbox commented:
The effect looks awesome! I have few questions in addition to the few earlier comments.

The transition logic doesn't seem to take elevation into account when transitioning to the mercator space. Do you think it's something we should consider adding later on if the zoom level for transition becomes adjustable (or someone uses a really high exaggeration)?

mpulkki-mapbox commented:
I also noticed that the altitude adjustment is absent in this PR. It means that the camera altitude is no longer converted correctly from the mercator space when the globe is enabled. Previously it could be seen as camera zooming in when moving towards polar regions as conversion between mercator projection and globe is not linear. That change was added in https://github.com/mapbox/mapbox-gl-js/pull/10984.

karimnaaji commented:
>1. The transition logic doesn't seem to take elevation into account when transitioning to the mercator space. Do you think it's something we should consider adding later on if the zoom level for transition becomes adjustable (or someone uses a really high exaggeration)?

Yes I think we should look into adding that. I haven't looked at elevation during the transition yet, but if we could use the up vector globe view vertices we should be able to account for it, as it is more noticeable with high exaggeration.

>Previously it could be seen as camera zooming in when moving towards polar regions

I looked into that this morning and I'm wondering whether that is overall a desired behavior. At a high level, this implies that the map feels like it's zooming in and out when we move along latitudes. Conceptually, this means that we keep the altitude conversions correct.

The end result feels a little surprising, even though it is conceptually correct. It seems like we would more naturally expect a constant globe sphere size rather than a change of camera altitude for that panning user interaction.

Is there any possible way forward where we keep constant sphere size (e.g. current behavior     of `pixelsPerMeter = mercatorZfromAltitude(1, lat) * worldSize` instead of `pixelsPerMeter = mercatorZfromAltitude(1, 0) * worldSize`) while correctly converting these altitude values as we move along latitudes? And do you foresee any issue with that approach?

karimnaaji commented:
>>The transition logic doesn't seem to take elevation into account when transitioning to the mercator space. Do you think it's something we should consider adding later on if the zoom level for transition becomes adjustable (or someone uses a really high exaggeration)?
>
>Yes I think we should look into adding that. I haven't looked at elevation during the transition yet, but if we could use the up vector globe view vertices we should be able to account for it, as it is more noticeable with high exaggeration.

I updated the PR to add support for elevation during the transition when using exaggeration in https://github.com/mapbox/mapbox-gl-js/pull/10999/commits/11c10220f87b4a2a283c42eace0533224fe0d14c.

https://user-images.githubusercontent.com/7061573/132929009-479bcd0f-c16c-43f6-8557-c2ac5fec59fd.mov

>2. What do you think we should do with line aligned labels? :)

Updated to fade out viewport aligned symbols in https://github.com/mapbox/mapbox-gl-js/pull/10999/commits/e91e6a1372e3211ebb864997c99c16411523a047 and https://github.com/mapbox/mapbox-gl-js/pull/10999/commits/34226d5af84398e46eb66e0107d66663486e50f8.

mpulkki-mapbox commented:
> I looked into that this morning and I'm wondering whether that is overall a desired behavior. Generally, I'm in favor of keeping this behavior and fixing the altitude conversions. At a high level, this implies that the map feels like it's zooming in and out when we pan along latitudes. Conceptually, this means that we keep the altitude conversions correct.
> 
> The end result for the user feels a little surprising, even though it is conceptually correct for altitude. One might expect the more natural behavior of a constant globe sphere size rather than a change of camera altitude for a panning interaction.
> 
> Is there any possible way forward where we keep constant sphere size (e.g. current behavior of `pixelsPerMeter = mercatorZfromAltitude(1, lat) * worldSize` instead of `pixelsPerMeter = mercatorZfromAltitude(1, 0) * worldSize`) while correctly converting these altitude values as we move along latitudes? And do you foresee any issue with that approach?
> 
> Another negative aspect of the implicit behavior of zooming in towards the pole is that we create more magnification artifacts during texture sampling as the draping was created for a texture intended to be displayed further away than where we'd place the camera at. This is something rather noticeable on features that may need more pixel accuracy such as the line boundaries here:

I see two ways forward with the behavior. Either:
 1. Keep zoom constant and compensate the camera altitude properly as it's currently being done
 2. Update zoom level in order to keep the altitude constant.

I don't remember if there was any reasoning behind the decision I initially made for the behavior. Texture magnification is indeed a con of this approach but on the other hand there's no tile reparsing nor additional tile requests made when panning the map. I believe both approaches have been used successfully by other globe implementations.
